README.md
[![Build Status](https://travis-ci.org/pine/p5-HTTP-Command-Wrapper.svg?branch=master)](https://travis-ci.org/pine/p5-HTTP-Command-Wrapper) [![Build Status](https://img.shields.io/appveyor/ci/pine/p5-HTTP-Command-Wrapper/master.svg)](https://ci.appveyor.com/project/pine/p5-HTTP-Command-Wrapper/branch/master) [![Coverage Status](http://codecov.io/github/pine/p5-HTTP-Command-Wrapper/coverage.svg?branch=master)](https://codecov.io/github/pine/p5-HTTP-Command-Wrapper?branch=master)
# NAME

HTTP::Command::Wrapper - The command based HTTP client (wget/curl wrapper). Too minimum dependencies!

# SYNOPSIS

    use HTTP::Command::Wrapper;

    my $client  = HTTP::Command::Wrapper->create; # auto detecting (curl or wget)
    my $content = $client->fetch('https://github.com/');

    print "$content\n";

# DESCRIPTION

HTTP::Command::Wrapper is a very simple HTTP client module.
It can wrap `wget` or `curl` command, and can use same interface.

# REQUIREMENTS

`wget` or `curl`

# METHODS

## CLASS METHODS

### `create($options = {})`

Create new wrapper instance using automatic commands detecting.

### `create($type, $options = {})`

Create new wrapper instance. `'wget'` or `'curl'` can be specified as `$type` value.

#### `$options`

- `verbose => 1`

    Turn on verbose output, with all the available data.

- `quiet => 1`

    Turn off output.

## METHODS

### `fetch($url, $headers = [])`

Fetch http/https contents from `$url`. Return a content body as string.

### `fetch_able($url, $headers = [])`

Return true if `$url` contents can fetch (status code is `200`).

### `download($url, $path, $headers = [])`

Fetch http/https contents from `$url`. Save in file. Return process exit code as boolean.

# LICENSE

The MIT License (MIT)

Copyright (c) 2015-2016 Pine Mizune

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

# AUTHOR

Pine Mizune &lt;pinemz@gmail.com>


Changes
Revision history for Perl extension HTTP-Command-Wrapper

{{$NEXT}}

0.08 2017-04-15T06:29:43Z

    - fix tests

0.07 2016-05-06T20:47:34Z

    - fix repository urls and badge urls

0.06 2016-01-23T22:20:21Z

    - support verbose and quiet options

0.05 2015-12-22T03:31:16Z

    - update document
    - change required perl version (5.8.2 -> 5.8.8)

0.04 2015-12-17T14:39:09Z

    - fix tests

0.03 2015-12-17T14:30:37Z

    - fix tests

0.02 2015-12-16T17:55:10Z

    - update document

0.01 2015-12-16T17:25:42Z

    - original version



t/Util.pm
use strict;
use warnings FATAL => 'all';
use utf8;

use Exporter qw/import/;

use File::Basename qw/dirname/;
use File::Spec;
use lib (
    File::Spec->catfile(dirname(__FILE__), qw/.. lib/),
    File::Spec->catfile(dirname(__FILE__), qw/lib/),
);

use Test::More 0.98;
use Test::Exception;

our @EXPORT = (
    @Test::More::EXPORT,
    @Test::Exception::EXPORT,
);

1;


t/lib/HTTP/Command/Wrapper/Test/Mock.pm
package HTTP::Command::Wrapper::Test::Mock;
use strict;
use warnings;
use utf8;

use Config;
use Cwd qw/abs_path/;
use Data::Section::Simple qw/get_data_section/;
use Exporter qw/import/;
use File::Basename qw/dirname/;
use File::Slurp qw/write_file/;
use File::Spec;
use Sys::Info;

our @EXPORT = qw/
    create_binary_mock
    create_dummy_wgetrc
    create_dummy_curlrc
/;

my $data_path = abs_path(
    File::Spec->catfile(dirname(__FILE__), qw/.. .. .. .. .. data/));

sub create_binary_mock (&) {
    my $code      = shift;
    my $mock_path = File::Spec->catfile($data_path, 'bin');
    my $path      = $mock_path;
    my $info      = Sys::Info->new;

    if ($info->os->is_windows) {
        $path .= $Config{path_sep};
        $path .= dirname($ENV{COMSPEC});
    }

    else {
        $path .= $Config{path_sep};
        $path .= $ENV{PATH};
    }

    local $ENV{PATH} = $path;
    return $code->();
}

sub create_dummy_wgetrc (&) {
    my $code      = shift;
    my $mock_path = File::Spec->catfile($data_path, 'rc/.wgetrc');

    local $ENV{WGETRC} = $mock_path;
    return $code->();
}

sub create_dummy_curlrc (&) {
    my $code = shift;

    my $home_path   = File::Spec->catfile($data_path, 'rc/');
    my $curlrc_path = File::Spec->catfile($data_path, 'rc/.curlrc');
    my $netrc_path  = File::Spec->catfile($data_path, 'rc/.netrc');

    my $curlrc_data = get_data_section('.curlrc');
    write_file($curlrc_path, sprintf($curlrc_data, $netrc_path));

    local $ENV{CURL_HOME} = $home_path;
    return $code->();
}

1;
__DATA__
@@ .curlrc
noproxy = "127.0.0.1"
netrc-file = "%s"


t/lib/HTTP/Command/Wrapper/Test/Server.pm
package HTTP::Command::Wrapper::Test::Server;
use strict;
use warnings;
use utf8;

use Cwd qw/abs_path/;
use File::Spec;
use File::Basename qw/dirname/;
use Test::TCP;
use Plack::Loader;
use Plack::Middleware::Static;

sub import {
    server();

    no strict 'refs';

    my $caller  = caller;
    *{"${caller}::$_"} = \&{__PACKAGE__."::$_"} for qw/uri_for path_for/;
}

my $server;
sub server {
    return $server if $server;

    my $app = sub {
        my $env = shift;

        Plack::Middleware::Static->new({
            path => sub { 1 },
            root => path_for(),
        })->call($env);
    };

    $server = Test::TCP->new(
        code => sub {
            my $port   = shift;
            my $server = Plack::Loader->auto(
                port => $port,
                host => '127.0.0.1',
            );
            $server->run($app);
        },
    );

    return $server;
}

sub port {
    return server()->port;
}

sub uri_for {
    my $path = shift;
    my $port = server()->port;
    return "http://127.0.0.1:$port/$path";
}

sub path_for {
    my $path = shift;
    $path = '' unless defined $path;
    return abs_path(File::Spec->catfile(dirname(__FILE__), qw/.. .. .. .. .. data htdocs/, $path)),
}

1;


lib/HTTP/Command/Wrapper/Curl.pm
package HTTP::Command::Wrapper::Curl;
use strict;
use warnings;
use utf8;

sub new {
    my ($class, $opt) = @_;
    return bless { opt => $opt } => $class;
}

sub fetch_able {
    my ($self, $url, $headers) = @_;

    my $command = $self->_build($headers, 1, [ '--head', qq/"$url"/ ]);
    `$command` =~ m/200 OK/;
}

sub fetch {
    my ($self, $url, $headers) = @_;

    my $command = $self->_build($headers, 1, [ qq/"$url"/ ]);
    `$command`;
}

sub download {
    my ($self, $url, $path, $headers) = @_;

    my $command = $self->_build($headers, 0, [ qq/"$url"/, '-o', qq/"$path"/ ]);
    system($command) == 0;
}

sub _build {
    my ($self, $headers, $quiet, $opts) = @_;
    my @args = (
        'curl',
        '-L',
        $self->_headers($headers),
        $quiet        ? undef      : $self->_verbose,
        $quiet        ? '--silent' : $self->_quiet,
        defined $opts ? @$opts     : undef,
    );
    return join(' ', grep { $_ } @args);
}

sub _headers {
    my ($self, $headers) = @_;
    return unless defined $headers;
    return if @$headers == 0;
    return join(' ', map { "-H \"$_\"" } @$headers);
}

sub _verbose {
    my $self = shift;
    return $self->{opt}->{verbose} ? '--verbose' : '';
}

sub _quiet {
    my $self = shift;
    return $self->{opt}->{quiet} ? '--silent' : '';
}

1;


lib/HTTP/Command/Wrapper/Wget.pm
package HTTP::Command::Wrapper::Wget;
use strict;
use warnings;
use utf8;

sub new {
    my ($class, $opt) = @_;
    return bless { opt => $opt } => $class;
}

sub fetch_able {
    my ($self, $url, $headers) = @_;

    my $command = $self->_build($headers, 1, [ '--server-response', '--spider', qq/"$url"/ ]);
    `$command 2>&1` =~ m/200 OK/;
}

sub fetch {
    my ($self, $url, $headers) = @_;

    my $command = $self->_build($headers, 1, [ qq/"$url"/, '-O', '-' ]);
    `$command`;
}

sub download {
    my ($self, $url, $path, $headers) = @_;

    my $command = $self->_build($headers, 0, [ '--continue', qq/"$url"/, '-O', qq/"$path"/ ]);
    system($command) == 0;
}

sub _build {
    my ($self, $headers, $quiet, $opts) = @_;
    my @args = (
        'wget',
        $self->_headers($headers),
        $quiet ? () : $self->_verbose,
        $quiet ? '--quiet' : $self->_quiet,
        @$opts
    );
    return join(' ', @args);
}

sub _headers {
    my ($self, $headers) = @_;
    $headers = [] unless defined $headers;

    return join(' ', map { "--header=\"$_\"" } @$headers);
}

sub _verbose {
    my $self = shift;
    return $self->{opt}->{verbose} ? '--verbose' : '';
}

sub _quiet {
    my $self = shift;
    return $self->{opt}->{quiet} ? '--quiet' : '';
}

1;


lib/HTTP/Command/Wrapper.pm
package HTTP::Command::Wrapper;
use strict;
use warnings;
use utf8;

our $VERSION = "0.07";

use File::Which qw/which/;

use HTTP::Command::Wrapper::Curl;
use HTTP::Command::Wrapper::Wget;

sub create {
    my $class = shift;
    my $opt   = {};
    my $type  = undef;

    $opt = $_[0] if ref $_[0] eq 'HASH';
    $opt = $_[1] if ref $_[1] eq 'HASH';

    $type = $_[0] unless ref $_[0];
    $type = $class->_detect_type unless defined $type;

    return HTTP::Command::Wrapper::Curl->new($opt) if $type eq 'curl';
    return HTTP::Command::Wrapper::Wget->new($opt) if $type eq 'wget';

    die 'Command not detected (curl or wget)';
}

sub _detect_type {
    my $class = shift;

    return 'curl' if $class->_which('curl');
    return 'wget' if $class->_which('wget');
    return undef;
}

sub _which { which($_[1]) }

1;
__END__

=encoding utf-8

=head1 NAME

HTTP::Command::Wrapper - The command based HTTP client (wget/curl wrapper). Too minimum dependencies!

=head1 SYNOPSIS

    use HTTP::Command::Wrapper;

    my $client  = HTTP::Command::Wrapper->create; # auto detecting (curl or wget)
    my $content = $client->fetch('https://github.com/');

    print "$content\n";

=head1 DESCRIPTION

HTTP::Command::Wrapper is a very simple HTTP client module.
It can wrap C<wget> or C<curl> command, and can use same interface.

=head1 REQUIREMENTS

C<wget> or C<curl>

=head1 METHODS

=head2 CLASS METHODS

=head3 C<create($options = {})>

Create new wrapper instance using automatic commands detecting.

=head3 C<create($type, $options = {})>

Create new wrapper instance. C<'wget'> or C<'curl'> can be specified as C<$type> value.

=head4 C<$options>

=over

=item C<verbose =E<gt> 1>

Turn on verbose output, with all the available data.

=item C<quiet =E<gt> 1>

Turn off output.

=back

=head2 METHODS

=head3 C<fetch($url, $headers = [])>

Fetch http/https contents from C<$url>. Return a content body as string.

=head3 C<fetch_able($url, $headers = [])>

Return true if C<$url> contents can fetch (status code is C<200>).

=head3 C<download($url, $path, $headers = [])>

Fetch http/https contents from C<$url>. Save in file. Return process exit code as boolean.

=head1 LICENSE

The MIT License (MIT)

Copyright (c) 2015-2016 Pine Mizune

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

=head1 AUTHOR

Pine Mizune E<lt>pinemz@gmail.comE<gt>

=cut

